### PR TITLE
enhance(dataset): tune processed keys for data loader

### PR DESCRIPTION
**Pull request by tianweidut**
## Modules
- [x] Python-SDK

## Checklist
- [x] run code format and lint check
- [x] add unit test
- [ ] add necessary doc

### CODE REVIEW

File: client/starwhale/api/_impl/dataset/loader.py [hunk 1]
```diff
@@ -172,20 +173,81 @@ def __init__(
...
         else:
             while True:
-                # TODO: tune last processed range for multithread
-                pk = [self.last_processed_range] if self.last_processed_range else None
+                pk = self._get_processed_key_range()
                 rt = self.session_consumption.get_scan_range(pk)
-                self.last_processed_range = rt
                 if rt is None:
-                    break
+                    if self._iter_row_processing or not self._check_pipe_data_empty():
```
Comment (goldenxinxing) [inline client/starwhale/api/_impl/dataset/loader.py:244]: Add a lock to ensure that the queues will not be updated during the check time?

File: client/starwhale/api/_impl/dataset/loader.py [hunk 2]
```diff
@@ -203,23 +265,32 @@ def _iter_meta(self) -> t.Generator[TabularDatasetRow, None, None]:
...
             for meta in self._iter_meta():
                 if meta and isinstance(meta, TabularDatasetRow):
-                    mq.put(meta)
+                    out_mq.put(meta)
         except Exception as e:
-            mq.put(e)
+            out_mq.put(e)
             raise
 
         for _ in range(0, self._num_workers):
```
Comment (goldenxinxing) [inline client/starwhale/api/_impl/dataset/loader.py:292]: use out_mq.task_done() to replace foreach? the same with function of '_unpack_row_for_queue'

File: client/starwhale/api/_impl/dataset/loader.py [hunk 3]
```diff
@@ -203,23 +265,32 @@ def _iter_meta(self) -> t.Generator[TabularDatasetRow, None, None]:
...
+
+    def _iter_meta_for_queue(self) -> None:
+        out_mq = self._meta_fetched_queue
+        if out_mq is None:
+            raise RuntimeError("queue not initialized for iter meta")
         try:
             for meta in self._iter_meta():
                 if meta and isinstance(meta, TabularDatasetRow):
-                    mq.put(meta)
+                    out_mq.put(meta)
```
[the remark this answers is not in the document]
Comment (tianweidut) [reply client/starwhale/api/_impl/dataset/loader.py:287]: fixed

File: client/starwhale/api/_impl/dataset/loader.py (hunk 1 above)
Comment (tianweidut) [reply client/starwhale/api/_impl/dataset/loader.py]: updated

**Review by goldenxinxing — APPROVED** [review]
LGTM

**Pull request merged**